packages/flutter_tools/lib/src/artifacts.dart

```diff
@@ -214,6 +203,48 @@ class ArtifactStore {
     _logging.fine('Wrote file');
   }
 
+  /// Download the artifacts.zip archive for the given platform from GCS
+  /// and extract it to the local cache.
+  static Future _doDownloadArtifactsFromZip(String platform) async {
+    String url = getCloudStorageBaseUrl(platform) + 'artifacts.zip';
+    _logging.info('Downloading $url.');
+
+    HttpClient httpClient = new HttpClient();
+    HttpClientRequest request = await httpClient.getUrl(Uri.parse(url));
+    HttpClientResponse response = await request.close();
+    _logging.fine('Received response statusCode=${response.statusCode}');
+    if (response.statusCode != 200)
+      throw new Exception(response.reasonPhrase);
+
+    BytesBuilder responseBody = new BytesBuilder(copy: false);
+    await for (List<int> chunk in response) {
+      responseBody.add(chunk);
+    }
+
+    Archive archive = new ZipDecoder().decodeBytes(responseBody.takeBytes());
+    Directory cacheDir = _getCacheDirForPlatform(platform);
+    for (ArchiveFile archiveFile in archive) {
+      File cacheFile = new File(path.join(cacheDir.path, archiveFile.name));
+      IOSink sink = cacheFile.openWrite();
+      sink.add(archiveFile.content);
+      sink.close();
+    }
+  }
+
+  /// A wrapper ensuring that a platform's ZIP is not downloaded multiple times
+  /// concurrently.
+  static Future _downloadArtifactsFromZip(String platform) {
+    if (_pendingZipDownloads.containsKey(platform)) {
+      return _pendingZipDownloads[platform].then((dynamic) {});
+    }
+    print('Downloading $platform artifacts from the cloud, one moment please...');
+    Future future = _doDownloadArtifactsFromZip(platform);
+    _pendingZipDownloads[platform] = future;
+    return future.then((dynamic) => _pendingZipDownloads.remove(platform));
+  }
+
+  static Map<String, Future> _pendingZipDownloads = {};
```

Comment:
final

Also, I'd initialize it to `new Map<String Future>` so that you have the right type at runtime.